crates/bevy_animation/src/graph.rs

```diff
@@ -0,0 +1,467 @@
+
+#[derive(Component, Reflect)]
+#[reflect(Component)]
+pub struct AnimationGraph {
+    graph: StableDiGraph<AnimationNode, ()>,
+    root: AnimationNodeIndex,
+}
+
+#[derive(Clone, Copy, Deref, Reflect)]
+pub struct AnimationWeight(f32);
+
+#[derive(Clone, Reflect)]
+pub struct AnimationNode {
+    animation: Option<PlayingAnimation>,
+    pub paused: bool,
+    pub weight: AnimationWeight,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Deref, Reflect)]
+pub struct AnimationNodeIndex(NodeIndex<u32>);
+
+struct AnimationContext<
+    'gr,
+    't,
+    'ac,
+    'p,
+    'ppq,
+    'ppa,
+    'ppb,
+    'ppc,
+    'cq,
+    'ca,
+    'cb,
+    'cc,
+    'nq,
+    'na,
+    'nb,
+    'nc,
+    'tq,
+    'ta,
+    'tb,
+    'tc,
+    'mwq,
+    'mwa,
+    'mwb,
+    'mwc,
+> {
+    animation_graph: &'gr mut AnimationGraph,
+    time: &'t Time,
+    root_entity: Entity,
+    animation_clips: &'ac Assets<AnimationClip>,
+    parent: Option<&'p Parent>,
+    parents: &'ppq Query<
+        'ppa,
+        'ppb,
+        (
+            Has<AnimationPlayer>,
+            Has<AnimationGraph>,
+            Option<&'ppc Parent>,
+        ),
+    >,
+    children: &'cq Query<'ca, 'cb, &'cc Children>,
+
+    // TODO: These should be abstracted into a list of `Animatable` properties.
+    names: &'nq Query<'na, 'nb, &'nc Name>,
+    transforms: &'tq Query<'ta, 'tb, &'tc mut Transform>,
+    morph_weights: &'mwq Query<'mwa, 'mwb, &'mwc mut MorphWeights>,
+}
+
+impl<
+        'gr,
+        't,
+        'ac,
+        'p,
+        'ppq,
+        'ppa,
+        'ppb,
+        'ppc,
+        'cq,
+        'ca,
+        'cb,
+        'cc,
+        'nq,
+        'na,
+        'nb,
+        'nc,
+        'tq,
+        'ta,
+        'tb,
+        'tc,
+        'mwq,
+        'mwa,
+        'mwb,
+        'mwc,
+    >
+    AnimationContext<
+        'gr,
+        't,
+        'ac,
+        'p,
+        'ppq,
+        'ppa,
+        'ppb,
+        'ppc,
+        'cq,
+        'ca,
+        'cb,
+        'cc,
+        'nq,
+        'na,
+        'nb,
+        'nc,
+        'tq,
+        'ta,
+        'tb,
+        'tc,
+        'mwq,
+        'mwa,
+        'mwb,
+        'mwc,
+    >
+{
+    fn evaluate(&mut self) {
+        self.evaluate_node(self.animation_graph.root, AnimationWeight(1.0))
+    }
+
+    fn evaluate_node(&mut self, node_index: AnimationNodeIndex, weight: AnimationWeight) {
+        let kids: Vec<_> = self
+            .animation_graph
+            .graph
+            .neighbors_directed(*node_index, Direction::Outgoing)
+            .collect();
+        for &kid in &kids {
+            // FIXME: Is `weight * kid_edge.weight()` right?
+            self.evaluate_node(kid.into(), weight * self.animation_graph.graph[kid].weight);
+        }
+
+        let node = &mut self.animation_graph.graph[*node_index];
+        if let Some(ref mut playing_animation) = node.animation {
+            crate::apply_animation(
```

Comment:
Each time `apply_animation` is called, the query for each bone within the hierarchy is redone, which involves a hierarchy traversal and multiple string comparisons. This works, but the total overhead scales with the number of clips in the graph. This isn't a significant departure from what we currently have, so I don't want to block on this, but we need to address this sooner rather than later.